static/app/components/structuredEventData/useExpandedState.tsx

```diff
@@ -0,0 +1,66 @@
+import type {ReactNode} from 'react';
+import {createContext, useCallback, useContext, useMemo, useRef} from 'react';
+
+import {uniq} from 'sentry/utils/array/uniq';
+
+type State = 'expanded' | 'collapsed';
+
+const context = createContext<{
+  collapse: (path: string) => void;
+  expand: (path: string) => void;
+  expandedPaths: string[];
+}>({
+  collapse: () => {},
+  expand: () => {},
+  expandedPaths: [],
+});
+
+interface Props {
+  children: ReactNode;
+  initialExpandedPaths: () => string[];
+  onToggleExpand?: (expandedPaths: string[], path: string, state: State) => void;
+}
+
+export function ExpandedStateContextProvider({
+  children,
+  initialExpandedPaths,
+  onToggleExpand,
+}: Props) {
+  const expandedRef = useRef<string[]>(initialExpandedPaths());
```

Comment:
Is there a reason we need to keep this `expandedRef` up to date when expanding/collapsing? It seems a bit confusing to me if we have two sources of truth in the component. Could we just use it for the initial state and leave it be after that?

Reply:
yes! it makes the callback api so much easier.

The main use-case for the callback is so we can keep track of everything that's expanded, and then re-render with that same state in the future. This is especially needed in breadcrumbs inside the replay details page where we use a virtual list, as the user scrolls list items are being unmounted and re-mounted all the time.

Without the callback a callsite would need to do the `.concat` and `.filter` calls (just a few lines below here). But with the ref the callback takes that on instead. Then we can shove the whole state into another ref that contains a list of states for our virtual list to use.

Reply:
Ahh okay, I see the utility now 👍